compiler/rustc_metadata/src/creader.rs

```diff
@@ -507,18 +508,31 @@ impl<'a> CrateLoader<'a> {
         }))
     }
 
-    fn resolve_crate<'b>(
+    fn resolve_crate_or_abort<'b>(
         &'b mut self,
         name: Symbol,
         span: Span,
         dep_kind: CrateDepKind,
     ) -> CrateNum {
+        self.resolve_crate(name, span, dep_kind).unwrap_or_else(|| FatalError.raise())
+    }
+
+    fn resolve_crate<'b>(
+        &'b mut self,
+        name: Symbol,
+        span: Span,
+        dep_kind: CrateDepKind,
+    ) -> Option<CrateNum> {
         self.used_extern_options.insert(name);
-        self.maybe_resolve_crate(name, dep_kind, None).unwrap_or_else(|err| {
-            let missing_core =
-                self.maybe_resolve_crate(sym::core, CrateDepKind::Explicit, None).is_err();
-            err.report(&self.sess, span, missing_core)
-        })
+        self.maybe_resolve_crate(name, dep_kind, None).map_or_else(
```

Comment:
```suggestion
        self.maybe_resolve_crate(name, dep_kind, None).or_else(
```
Just `or_else` without `map` looks like the right fit here.

Reply:
The return type of `CrateLoader::maybe_resolve_crate` is `Result<CrateNum, CrateError>`, but the target return type that is meant to be propagated to `CrateLoader::resolve_crate` is an `Option<CrateNum>`. `or_else` is not going to work in this case because either I'd like to map a `Ok(cnum)` to a `Some(cnum)` or in case of `Err(err)` I would like to unpack the error, emit it, and return `None`. An alternative that could work, would be
```rust
self.maybe_resolve_crate(name, dep_kind, None).or_else(|err| {
    let missing_core =
        self.maybe_resolve_crate(sym::core, CrateDepKind::Explicit, None)
            .is_err();
        err.report(&self.sess, span, missing_core);
        err
    }
).ok()
```
But to make this work I would have to change `CrateError::report(self, ...)` to `CrateError::report(&self, ...)`, which I personally would find just a bit less desirable, given that semantically I think reporting an error "should" consume it.

Or the cleanest solution may in fact be a simple `match`.